plugins/weida-raw/fu-weida-raw-device.c

```diff
@@ -0,0 +1,1093 @@
+
+	return fu_weida_raw_w8760_command_read(self, cmd, sizeof(cmd), buf, 0, 10, error);
+}
+
+static guint8
+fu_weida_raw_w8760_get_device_mode(FuWeidaRawDevice *self, GError **error)
+{
+	guint8 status[10] = {0};
+	if (!self)
+		return 0;
+
+	if (fu_weida_raw_w8760_get_status(self, status, 4, 1, error)) {
+		g_debug("status = %d", status[0]);
+		return status[0];
+	}
+
+	return 0;
+}
+
+static gboolean
+fu_weida_raw_w8760_protect_flash(FuWeidaRawDevice *self, guint16 protect_mask, GError **error)
+{
+	guint16 mask = protect_mask;
+
+	guint8 cmd[10] = {FU_WDT8760_COMMAND9, FU_WDT8760_PROTECT_FLASH, 0, 0, 0, 0, 0};
+	fu_memwrite_uint16(&cmd[2], mask, G_LITTLE_ENDIAN);
+	fu_memwrite_uint16(&cmd[4], ~mask, G_LITTLE_ENDIAN);
+
+	if (!fu_weida_raw_device_set_feature(self, cmd, sizeof(cmd), error))
+		return FALSE;
+	else
+		return TRUE;
+}
+
+static gboolean
+fu_weida_w8760_set_n_check_device_mode(FuWeidaRawDevice *pdev,
+				       guint8 mode,
+				       int timeout_ms,
+				       int intv_ms,
+				       GError **error)
+{
+	int polling_timeout_ms = 300 * 1000;
+	int polling_intv_ms = 30;
+
+	if (timeout_ms)
+		polling_timeout_ms = timeout_ms;
+	if (intv_ms)
+		polling_intv_ms = intv_ms;
+
+	do {
+		if (fu_weida_raw_w8760_set_device_mode(pdev, mode, error)) {
+			if (fu_weida_raw_w8760_get_device_mode(pdev, error) == mode)
+				return TRUE;
+		}
+
+		g_usleep(polling_intv_ms * 1000);
+		polling_timeout_ms -= polling_intv_ms;
+	} while (polling_timeout_ms > 0);
+
+	return FALSE;
+}
+
+static gboolean
+fu_weida_raw_w8760_wait_cmd_end(FuWeidaRawDevice *self, int timeout_ms, int invt_ms, GError **error)
+{
+	int polling_timeout_ms = 5000;
+	int polling_intv_ms = 30;
+	int status;
+	guint8 status_buf[10] = {0};
+
+	if (timeout_ms)
+		polling_timeout_ms = timeout_ms;
+	if (invt_ms)
+		polling_intv_ms = invt_ms;
+
+	do {
+		if (fu_weida_raw_w8760_get_status(self, &status_buf[0], 0, 1, error) <= 0) {
+			g_debug("fail %s \n", __func__);
+			return FALSE;
+		}
+
+		status = status_buf[0];
+		if ((status & 0x01) == 0) {
+			return TRUE;
+		}
+
+		fu_device_sleep(FU_DEVICE(self), polling_intv_ms);
+		polling_timeout_ms -= polling_intv_ms;
+	} while (polling_timeout_ms > 0);
+	g_debug("wait_cmd_end timeout fail");
+
+	return FALSE;
+}
+
+static gboolean
+fu_weida_raw_w8760_erase_flash(FuWeidaRawDevice *self,
+			       guint32 address,
+			       guint32 size,
+			       GError **error)
+{
+	guint8 cmd[10] = {0};
+
+	cmd[0] = FU_WDT8760_COMMAND9;
+	cmd[1] = FU_WDT8760_ERASE_FLASH;
+	cmd[2] = (guint8)(address >> 12);
+	cmd[3] = (guint8)(((address & 0x0FFF) + size + 4095) >> 12);
+
+	if (!fu_weida_raw_device_set_feature(self, cmd, sizeof(cmd), error)) {
+		return FALSE;
+	} else {
+		if (!fu_weida_raw_w8760_wait_cmd_end(self, 0, 0, error)) {
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_w8760_set_flash_address(FuWeidaRawDevice *self, guint32 address, GError **error)
+{
+	guint8 cmd[10] = {0};
+	memset(&cmd, 0, sizeof(cmd));
```

Comment:
you don;t need the memset if you used {0}

Reply:
thanks solved